File: language-support/js/daml-ledger/src/mapping/create_command.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as mapping from '.';
import * as ledger from '..';
import * as grpc from 'daml-grpc';

export const CreateCommand: mapping.Mapping<grpc.CreateCommand, ledger.CreateCommand> = {
    toObject(command: grpc.CreateCommand): ledger.CreateCommand {
        return {
            templateId: mapping.Identifier.toObject(command.getTemplateId()!),
            arguments:  mapping.Record.toObject(command.getCreateArguments()!)    
        }
    },
    toMessage(command: ledger.CreateCommand): grpc.CreateCommand {
        const result = new grpc.CreateCommand();
        result.setTemplateId(mapping.Identifier.toMessage(command.templateId));
        result.setCreateArguments(mapping.Record.toMessage(command.arguments));
        return result;
    }
}

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/CreateCommand.java
// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v1.CommandsOuterClass;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.Objects;

public class CreateCommand extends Command {

    private final Identifier templateId;

    private final Record createArguments;

    public CreateCommand(@NonNull Identifier templateId, @NonNull Record createArguments) {
        this.templateId = templateId;
        this.createArguments = createArguments;
    }

    @Override
    public String toString() {
        return "CreateCommand{" +
                "templateId=" + templateId +
                ", createArguments=" + createArguments +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CreateCommand that = (CreateCommand) o;
        return Objects.equals(templateId, that.templateId) &&
                Objects.equals(createArguments, that.createArguments);
    }

    @Override
    public int hashCode() {

        return Objects.hash(templateId, createArguments);
    }

    @NonNull

    @Override
    public Identifier getTemplateId() {
        return templateId;
    }

    @NonNull
    public Record getCreateArguments() {
        return createArguments;
    }

    public static CreateCommand fromProto(CommandsOuterClass.CreateCommand create) {
        Record createArgument = Record.fromProto(create.getCreateArguments());
        Identifier templateId = Identifier.fromProto(create.getTemplateId());
        return new CreateCommand(templateId, createArgument);
    }

    public CommandsOuterClass.CreateCommand toProto() {
        return CommandsOuterClass.CreateCommand.newBuilder()
                .setTemplateId(this.templateId.toProto())
                .setCreateArguments(this.createArguments.toProtoRecord())
                .build();
    }
}
